### PR TITLE
fix: Add output of failed events to logs

### LINKED ISSUE #234: Specifying non-existing Service Account leads to job not starting
This was testedwith jes pre-release 0.2.0-next.0.

When specifying serviceAccount of a job, e.g.,

```
  - name: "Deploy using helm"
    events:
      - name: "sh.keptn.event.je-deployment.triggered"
    tasks:
      - name: "Run helm"
        files:
          - /charts
        env:
          - name: IMAGE
            value: "$.data.configurationChange.values.image"
            valueFrom: event
        image: "alpine/helm:3.7.2"
        serviceAccount: "jes-deploy-using-helm"
        cmd: ["helm"]
```

without `jes-deploy-using-helm` existing in the current namespace, the job is not starting (and eventually times out).

# Expected Behaviour

There should be an error sent back to Keptn (either as a .finished event with status fail/error, or a dedicated error log).

### LINKED ISSUE #235: Job that cannot start pod -> orphaned
When Job Executor Service creates a k8s Job that cannot spawn pod (if the wrong serviceAccount for the task is specified for example) the sequence fails because of timeout but no logs can be retrieved as there are no pods to fetch them from.
Furthermore the created job is not collected by the k8s TTL controller as it never finishes, so it will keep trying to spawn pods long after Job Executor Service gave up on it (possibly indefinitely if there's a configuration error) so it has to be manually removed by the user.

The Job Executor Service should detect that the job failed to start and add relevant information for the user extracted from the job status/events and explicitly delete the job if it didn't spawn any pods.

### How to reproduce:
Use a job config with a wrong service account:

```
apiVersion: v2
actions:
  - name: "Hello World e2e test"
    events:
      - name: "sh.keptn.event.deployment.triggered"
    tasks:
      - name: "Greet the world"
        image: "alpine"
        serviceAccount: "inexistentServiceAccount"
        cmd:
          - echo
        args:
          - "Hello World"
```

### PR DESCRIPTION
**Pull request by christian-kreuzberger-dtx**
Disclaimer: Thanks to the code provided by @Raffy23 in PR https://github.com/keptn-contrib/job-executor-service/pull/229

Fixes #234 by providing an appropriate error message

image

Unfortunately, we cannot easily fix #235, as it would require us to delete jobs (which we shouldn't do at the moment). I would dedicate the fix of #235 to the refactoring ticket of #244

### CODE REVIEW

**Comment by Raffy23** [comment]
Shouldn't we also implement a basic unit test for `GetFailedEventsForJob`?

**Comment by christian-kreuzberger-dtx** [comment]
Added two unit tests and rebased `main` branch

File: pkg/k8sutils/job_test.go [hunk 1]
```diff
@@ -740,6 +740,88 @@ func TestAwaitK8sJobDoneErrorJobFailed(t *testing.T) {
...
+			Conditions: []v1.JobCondition{}, // this job won't have a condition set
+		},
+	}
+
+	// two failed and one successful event
+	events := []corev1.Event{
+		{ObjectMeta: metav1.ObjectMeta{Name: "event-1"}, TypeMeta: metav1.TypeMeta{Kind: "job"}, InvolvedObject: corev1.ObjectReference{Name: jobName, Kind: "job", Namespace: namespace}, Source: corev1.EventSource{}, Reason: "FailedCreate", Message: failedEventMessage1},
+		{ObjectMeta: metav1.ObjectMeta{Name: "event-2"}, TypeMeta: metav1.TypeMeta{Kind: "job"}, InvolvedObject: corev1.ObjectReference{Name: jobName, Kind: "job", Namespace: namespace}, Reason: "Failed", Message: failedEventMessage2},
+		{ObjectMeta: metav1.ObjectMeta{Name: "event-3"}, TypeMeta: metav1.TypeMeta{Kind: "job"}, InvolvedObject: corev1.ObjectReference{Name: jobName, Kind: "job", Namespace: namespace}, Reason: "Pulled", Message: successfulEventMessage1},
+	}
```
Comment (christian-kreuzberger-dtx) [inline pkg/k8sutils/job_test.go:798]: I'm not 100% sure if what I'm doing here is correct - haven't had much experience with "faking" events in the fake client. Please help me out ^^

File: pkg/k8sutils/job_test.go [hunk 2]
```diff
@@ -740,6 +740,88 @@ func TestAwaitK8sJobDoneErrorJobFailed(t *testing.T) {
...
+		context.Background(), &job, metav1.CreateOptions{},
+	)
+
+	// create events
+	for _, event := range events {
+		_, err := k8sClientSet.CoreV1().Events(namespace).Create(
+			context.Background(), &event, metav1.CreateOptions{},
+		)
+
+		require.NoError(t, err)
```
Comment (christian-kreuzberger-dtx) [inline pkg/k8sutils/job_test.go:811]: Just want to check if adding the event works... sometimes it fails (e.g., duplicate name)

File: pkg/k8sutils/job_test.go [hunk 3]
```diff
@@ -740,6 +740,88 @@ func TestAwaitK8sJobDoneErrorJobFailed(t *testing.T) {
...
+	}
+
+	// create job
+	k8sClientSet.BatchV1().Jobs(namespace).Create(
+		context.Background(), &job, metav1.CreateOptions{},
+	)
+
+	// create events
+	for _, event := range events {
+		_, err := k8sClientSet.CoreV1().Events(namespace).Create(
```
Comment (christian-kreuzberger-dtx) [inline pkg/k8sutils/job_test.go:807]: Is it possible to add more than one event with a single function call?

File: pkg/k8sutils/job_test.go [hunk 4]
```diff
@@ -740,6 +740,88 @@ func TestAwaitK8sJobDoneErrorJobFailed(t *testing.T) {
...
+	successfulEventMessage1 := "Successfully pulled image"
+
+	// job config
+	job := v1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: jobName,
+		},
+		Spec: v1.JobSpec{},
+		Status: v1.JobStatus{
+			Conditions: []v1.JobCondition{}, // this job won't have a condition set
```
Comment (Raffy23) [inline pkg/k8sutils/job_test.go:789]: I'm not sure about the Kubernetes API at this point, but shouldn't the JobStatus be completely empty if the job hangs in at the point of not being able to create a single pod?

**Review by Raffy23 — APPROVED** [review]

File: pkg/k8sutils/job_test.go (hunk 3 above)
Comment (Raffy23) [reply pkg/k8sutils/job_test.go:807]: I don't think that the k8s API provides a convenient way to create multiple events, but I will let you know if I find a solution.

File: pkg/k8sutils/job_test.go (hunk 4 above)
Comment (christian-kreuzberger-dtx) [reply pkg/k8sutils/job_test.go:789]: Using it the same way as in other tests - I believe that's the correct approach

**Pull request merged**